src/core/ui/egl/rpi-input.cpp

```diff
@@ -0,0 +1,221 @@
+#if OTTO_UI_EGL
+
+#include <fcntl.h>
+#include <linux/input.h>
+#include <plog/Log.h>
+#include <string>
+#include <vector>
+#include "core/ui/mainui.hpp"
+#include "filesystem.hpp"
+
+namespace otto::ui {
+  static auto constexpr RELEASE = 0;
+  static auto constexpr PRESS   = 1;
+  static auto constexpr REPEAT  = 2;
+
+  namespace fs = otto::filesystem;
+
+  static bool ends_with(const std::string& s, const std::string& suffix)
+  {
+    return s.rfind(suffix) == (s.size() - suffix.size());
+  }
+
+  static int open_device(const std::string& device_type)
+  {
+    auto path = fs::path("/dev/input/by-id");
+
+    for (const auto& entry : fs::directory_iterator(path)) {
+      auto file = entry.path().string();
+      if (ends_with(file, device_type)) {
+        auto fullpath = path / file;
+        auto fd       = open(fullpath.c_str(), O_RDONLY | O_NONBLOCK);
+        auto result   = ioctl(fd, EVIOCGRAB, 1);
+        if (result != 0) {
+          LOGE << "Couldn't get exclusive input access to "
+               << fullpath.string();
+        }
+
+        return fd;
+      }
+    }
+
+    return -1;
+  }
+
+  std::unique_ptr<std::vector<input_event>> read_events(int device)
+  {
+    struct input_event events[64];
```

Comment:
is 64 a posix guarantied number here or a realistic max estimate? is there no macro for it? either way, a comment explaining it would be nice

Reply:
The 64 is arbitrary.  It's just "read a bunch of items from the input".  I'll move it to a named value, for self-documentation.